### PR TITLE
Validate maxWriteSize of a transaction before compression

### PR DESCRIPTION
**Pull request by SravanthiAshokKumar**
Limiting the LogData payload size after compression does not give a clear idea of the payload that is uncompressed as the compression ratio for each transaction data will be varied. This change avoids OOM issues in a client due to a single transaction exceeding the memory limits.

## Overview

Description:

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java [hunk 1]
```diff
@@ -193,6 +193,18 @@ public int getSizeEstimate() {
         return 1;
     }
 
+    private int getSizeEstimate(byte[] tempData) {
```
[the remark this answers is not in the document]
Comment (SravanthiAshokKumar) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java]: The PR was not up-to-date with the parallel PR for 412 branch. Updated now to latest.

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/ILogData.java [hunk 2]
```diff
@@ -52,9 +52,9 @@ public ILogData getSerialized() {
          * @param data     the log data to manage
          * @param metadata whether metadata needs to be serialized
          */
-        public SerializationHandle(ILogData data, boolean metadata) {
+        public SerializationHandle(ILogData data, boolean metadata, boolean checkSize, int limit) {
```
[the remark this answers is not in the document]
Comment (SravanthiAshokKumar) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/ILogData.java]: Done

**Review by Lujie1996 — APPROVED** [review]

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java [hunk 3]
```diff
@@ -382,24 +385,16 @@ public String toString() {
...
-            log.trace("checkMaxWriteSize: payload size is {} bytes.", payloadSize);
+    private void checkMaxWriteSizeIfRequired(Optional<Integer> limit, ByteBuf buf) {
+        if (!limit.isPresent()) {
+            return;
         }
 
-        if (payloadSize > limit) {
-            throw new WriteSizeException(payloadSize, limit);
-        }
+        int payloadSize = buf.readableBytes();
```
Comment (zfrenette) [inline runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java:393]: This seems correct for the existing code paths, but I think the computation would be more robust against future changes if the difference between the old and new writerIndex value is computed explicitly.

**Review by zfrenette — APPROVED** [review]
Looks good - Shared a few minor suggestions.

File: test/src/test/java/org/corfudb/runtime/checkpoint/CheckpointSmokeTest.java [hunk 4]
```diff
@@ -708,6 +705,7 @@ public void checkpointWriterSizeLimitViolationTest() {
 
         // set the max write size to 25 MB
         getRuntime().getParameters().setMaxWriteSize(maxWriteSize);
+        r.getParameters().setMaxWriteSize(maxWriteSize << 1);
```
[the remark this answers is not in the document]
Comment (SravanthiAshokKumar) [reply test/src/test/java/org/corfudb/runtime/checkpoint/CheckpointSmokeTest.java:708]: Sorry, this is misleading. This is not required with the latest changes.

**Review by Lujie1996 — APPROVED** [review]
Thank you for the tireless effort. The second commit with both compressed & uncompressed size checks looks good to me.

**Review by hisundar — APPROVED** [review]

**Pull request merged**